doc/testing_snitch.md

```diff
@@ -0,0 +1,146 @@
+ - [install tests](#install-tests) (in `test/install_tests`)
+
+
+### Run-time unit tests
+
+The purpose of these tests is to check the detailed run-time behavior of _snitch_. This includes testing all foreseeable edge cases, happy path and sad path, etc. These tests contribute to the test coverage analysis.
+
+The run-time unit tests are roughly organized by theme into separate files in the `tests/runtime_tests` directory. New files can be added there if necessary. There is no enforced mapping between test files and library source files. Likewise, there is no enforced mapping between test case names and tested classes / functions; this is left of the best judgment of the contributor.
+
+Run-time unit tests come in two flavor: using _doctest_ to test _snitch_, or using _snitch_ to test itself. The latter is less safe, since a bug in _snitch_ might accidentally make these tests pass (e.g., if the test cases are not running at all), but it usually provides nicer tests results on failure.
+
+The *test* code must be written using the _snitch_ "shorthand" API (e.g., `CHECK(...)`). Wrapper macros are used to wrap _doctest_ macros so the same API can be used in both flavors. The *tested* code must be written with the _snitch_ "prefixed" API (e.g., `SNITCH_CHECK(...)`), otherwise it would turn into _doctest_ code in the _doctest_-flavored run. For example:
+
+```c++
+// Use the shorthand TEST_CASE macro to create a new test
+// (this will use either snitch or doctest)
+TEST_CASE("test register a test case") {
+    snitch::registry r;
+
+    // Use prefixed macro for the code to be tested
+    // (this will always use snitch)
+    r.add({"a new test case", "[tag1][tag2]"}, SNITCH_CURRENT_LOCATION, []() {
+        SNITCH_CHECK(1 == 2);
+    });
+
+    // Use the shorthand CHECK macro to actually check something
+    // (this will use either snitch or doctest)
+    CHECK(r.test_cases.begin() != r.test_cases.end());
+}
+```
+
+
+### Compile-time unit tests
+
+Not implemented yet. The purpose of these tests will be to check how _snitch_ behaves at compile-time. In particular, to ensure that some specific usage will *not* compile. This cannot be achieved with the regular run-time tests, because it would prevent the whole test suite from compiling at all. Compile-time tests are better implemented at the build-system level, since the build system knows what compiler to use, what options, etc.
+
+
+### Approval tests
+
+These tests provide a more manageable way to check the output of _snitch_ for each reporter. Rather than use hand-written checks ("this line must contain this string, and then this other string"), which would be painful to write to achieve 100% coverage, the approach is to capture the output and save it to a file. If the output is satisfactory, we commit it as the new "expected" output, and future test runs just need to check that the output is compatible with the "expected" output. These tests contribute to the test coverage analysis.
```

Comment:
```suggestion
These tests provide a more manageable way to check the output of _snitch_ for each reporter. Rather than use handwritten checks ("this line must contain this string, and then this other string"), which would be painful to write to achieve 100% coverage, the approach is to capture the output and save it to a file. If the output is satisfactory, we commit it as the new "expected" output, and future test runs just need to check that the output is compatible with the "expected" output. These tests contribute to the test coverage analysis.
```
See https://github.com/snitch-org/snitch/pull/81#pullrequestreview-1380810467.